### PR TITLE
Adds missing description for usage in a laminas-mvc based application

### PR DESCRIPTION
**Pull request by froschdesign**

### CODE REVIEW

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+Create a listener as class using [a listener aggregate](../aggregates.md) and add the logger via constructor injection to create a valid object with all required dependencies, e.g. `module/Application/src/Listener/ErrorListener.php`:
+
+```php
+namespace Application\Listener;
+
+use Exception;
+use Laminas\EventManager\AbstractListenerAggregate;
+use Laminas\EventManager\EventManagerInterface;
+use Laminas\Log\Logger;
```
Comment (Ocramius) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md:21]: Should we endorse using a PSR logger? :thinking:

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+NOTE: **More Events**
+All events that can be triggered are listed and explained in the [documentation of laminas-mvc](https://docs.laminas.dev/laminas-mvc/mvc-event/).
+
+## Register Listener
+
+To register a listener in a laminas-mvc based application use the application or module configuration, such as `config/autload/global.php` or `module/Application/config/module.config.php`, and the configuration key `listeners`.
+
+The current example uses the module configuration, e.g. `module/Application/config/module.config.php`:
+
+<pre class="language-php" data-line="4-6"><code>
```
Comment (Ocramius) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md:62]: Is `pre` preferable, compared to triple-backticks? Or is this just for highlighting?

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+];
+</code></pre>
+
+All listeners registered in this way are fetched from the application service container.
+This means the listeners must be registered for the application service container.
+
+To register the listener for the application service container, extend the configuration of the module.
+Add the following lines to the module configuration file, e.g. `module/Application/config/module.config.php`:
+
+<pre class="language-php" data-line="3,8"><code>
```
Comment (Ocramius) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md:79]: This uses `data-line="3,8"`, while above it uses a range? Intentional? (just asking - unfamiliar with syntax in here)

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+a laminas-mvc based application.
+The example creates a listener for logging errors of an application with [laminas-log](https://docs.laminas.dev/laminas-log/).
+
+Before starting, make sure laminas-log is [installed and configured](https://docs.laminas.dev/laminas-log/installation/).
+
+laminas-eventmanager is already present in laminas-mvc based applications as it is an event-driven MVC layer based on the event manager.
+
+## Create Listener
+
+Create a listener as class using [a listener aggregate](../aggregates.md) and add the logger via constructor injection to create a valid object with all required dependencies, e.g. `module/Application/src/Listener/ErrorListener.php`:
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
Create [a listener aggregate](../aggregates.md) that defines and registers a listener, using a logger injected via its constructor; as an example, consider the following definition in the file `module/Application/src/Listener/ErrorListener.php`:
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md (hunk 1 above)
Comment (froschdesign) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md:21]: The laminas-log component enjoys automatic and seamless integration in a laminas-mvc based application via the laminas-component-installer, so I took these instead of a PSR logger.

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        if ($exception) {
+            $this->logger->crit('Error: ' . $exception->getMessage());
+        } else {
+            $this->logger->crit('Error: ' . $event->getError());
+        }
+    }
+}
+```
+
+The listener aggregate was chosen because [it can listen to even more events](../aggregates.md#recommendations) due to the possible extension.
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
The listener aggregate was chosen because [to allow it to listen to additional events](../aggregates.md#recommendations) if we want to later.
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+NOTE: **More Events**
+All events that can be triggered are listed and explained in the [documentation of laminas-mvc](https://docs.laminas.dev/laminas-mvc/mvc-event/).
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
All laminas-mvc events that can be triggered are listed and explained in the [laminas-mvc documentation](https://docs.laminas.dev/laminas-mvc/mvc-event/).
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+## Register Listener
+
+To register a listener in a laminas-mvc based application use the application or module configuration, such as `config/autload/global.php` or `module/Application/config/module.config.php`, and the configuration key `listeners`.
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
To register a listener in a laminas-mvc based application, use either application or module configuration (such as `config/autload/global.php` or `module/Application/config/module.config.php`, respectively), and define the configuration key `listeners`.
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+The current example uses the module configuration, e.g. `module/Application/config/module.config.php`:
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
This example uses module configuration, e.g. `module/Application/config/module.config.php`:
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+return [
+    'listeners' => [
+        Listener\ErrorListener::class,
+    ],
+    // …
+];
+</code></pre>
+
+All listeners registered in this way are fetched from the application service container.
+This means the listeners must be registered for the application service container.
```
Comment (weierophinney) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
This means the listeners must be registered for the application service container to discover and create them.
```

File: mkdocs.yml [hunk 10]
```diff
@@ -6,22 +6,24 @@ nav:
...
+        - LazyEventListener: lazy-listeners/lazy-event-listener.md
+        - LazyListenerAggregate: lazy-listeners/lazy-listener-aggregate.md
+    - "EventManager API": api.md
+    - "Intercepting Filters": intercepting-filters.md
+    - "Application Integration":
+        - "Usage in a laminas-mvc Application": application-integration/usage-in-a-laminas-mvc-application.md
+    - "Migration Guide":
+        - Introduction: migration/intro.md
+        - "Removed Functionality": migration/removed.md
+        - "Changed Functionality": migration/changed.md
```
Comment (weierophinney) [inline mkdocs.yml]: The promotion of these pages to a higher level in the TOC feels like an unrelated change; could we maybe do this part in a separate PR?

**Review by weierophinney — COMMENTED** [review]
Provided a number of suggestions; overall, structure and the example are fantastic!

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md (hunk 2 above)
Comment (froschdesign) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md:62]: > Or is this just for highlighting?

Correct, the Markdown extension does not allow / support to add `data-*` attributes which are used for line highlight.

Example: https://docs.laminas.dev/laminas-form/v3/application-integration/usage-in-a-laminas-mvc-application/#register-form-and-controller

Reference:

* https://prismjs.com/plugins/line-highlight/
* https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md033

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md (hunk 1 above)
Comment (Ocramius) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md:21]: Fair enough

Comment (froschdesign) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md:21]: The question is valid, because I have also been thinking about this topic for a while.

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md (hunk 3 above)
Comment (froschdesign) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md:79]: Lines can be highlighted as groups or individually, and there can be multiple highlights in one block of code.

File: mkdocs.yml (hunk 10 above)
Comment (froschdesign) [reply mkdocs.yml]: I have reverted the changes, but only a depth of two is correctly displayed in the navigation: https://docs.laminas.dev/laminas-eventmanager/

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        if ($exception) {
+            $this->logger->crit('Error: ' . $exception->getMessage());
+        } else {
+            $this->logger->crit('Error: ' . $event->getError());
+        }
+    }
+}
+```
+
+The listener aggregate was chosen because [to allow it to listen to additional events](../aggregates.md#recommendations) if we want to later.
```
Comment (gsteel) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]: Perhaps:
> The listener aggregate was chosen because it is capable of listening to multiple events, which could prove useful in future.

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+];
+</code></pre>
+
+The example uses the [reflection factory from laminas-servicemanager](https://docs.laminas.dev/laminas-servicemanager/reflection-abstract-factory/) to resolve the constructor dependencies for the listener class.
+
+If the listener has no dependencies, use the [factory `Laminas\ServiceManager\Factory\InvokableFactory`](https://docs.laminas.dev/laminas-servicemanager/configuring-the-service-manager/#factories).
+
+## Logging Message
+
+To log a message, produce a [dispatch error](https://docs.laminas.dev/laminas-mvc/mvc-event/#mvceventevent_dispatch_error-dispatcherror) by navigate to `http://localhost:8080/1234`, and the 404 page should be displayed.
```
Comment (gsteel) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
To log a message, produce a [dispatch error](https://docs.laminas.dev/laminas-mvc/mvc-event/#mvceventevent_dispatch_error-dispatcherror) by navigating to `http://localhost:8080/1234`, where a 404 page should be displayed.
```

File: doc/book/application-integration/usage-in-a-laminas-mvc-application.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+```
+
+The listener aggregate was chosen because [to allow it to listen to additional events](../aggregates.md#recommendations) if we want to later.
+
+NOTE: **More Events**
+All laminas-mvc events that can be triggered are listed and explained in the [laminas-mvc documentation](https://docs.laminas.dev/laminas-mvc/mvc-event/).
+
+## Register Listener
+
+To register a listener in a laminas-mvc based application, use either application or module configuration (such as `config/autload/global.php` or `module/Application/config/module.config.php`, respectively), and define the configuration key `listeners`.
```
Comment (gsteel) [inline doc/book/application-integration/usage-in-a-laminas-mvc-application.md]:
```suggestion
To register a listener in a laminas-mvc based application, use either application or module configuration (such as `config/autoload/global.php` or `module/Application/config/module.config.php`, respectively), and define the configuration key `listeners`.
```

Comment (froschdesign) [reply doc/book/application-integration/usage-in-a-laminas-mvc-application.md]: Good catch! 👍🏻

**Review by Ocramius — APPROVED** [review]
:ship: IMO :)

**Comment by Ocramius** [comment]
Thanks @froschdesign, @weierophinney, @gsteel!
